LEGO1/lego/legoomni/src/entity/legojetskiraceactor.cpp

```diff
@@ -11,17 +22,120 @@ LegoJetskiRaceActor::LegoJetskiRaceActor()
 	m_unk0x150 = 1.5f;
 }
 
-// STUB: LEGO1 0x10081120
+// FUNCTION: LEGO1 0x10081120
+// FUNCTION: BETA10 0x100ce19f
 MxS32 LegoJetskiRaceActor::VTable0x1c(LegoPathBoundary* p_boundary, LegoEdge* p_edge)
 {
-	// TODO
-	return 0;
+	// These are almost certainly not the correct names, but they produce the correct BETA10 stack
+	Mx3DPointFloat a;
+	Mx3DPointFloat bbb;
+	Mx3DPointFloat c;
+
+	// These names are verified by an assertion below
+	Vector3* v1 = NULL;
+	Vector3* v2 = NULL;
+
+	if (m_state == 1) {
+		if (m_destEdge == LegoPathController::GetControlEdgeA(13)) {
+			m_boundary = (LegoPathBoundary*) m_destEdge->OtherFace(LegoPathController::GetControlBoundaryA(13));
+		}
+		else if (m_destEdge == LegoPathController::GetControlEdgeA(15)) {
+			m_boundary = (LegoPathBoundary*) m_destEdge->OtherFace(LegoPathController::GetControlBoundaryA(15));
+		}
+
+		m_state = 0;
+		m_unk0x7c = 0;
+
+		if (m_userNavFlag) {
+			NavController()->SetLinearVel(m_worldSpeed);
+			return 0;
+		}
+		else {
+			return 1;
+		}
+	}
+	else {
+		if (p_edge == LegoPathController::GetControlEdgeA(12)) {
+			m_state = 1;
+
+			if (m_worldSpeed < g_eight) {
+				m_worldSpeed = g_eight;
+			}
+
+			m_destEdge = LegoPathController::GetControlEdgeA(13);
+			m_boundary = LegoPathController::GetControlBoundaryA(13);
+		}
+		else if (p_edge == LegoPathController::GetControlEdgeA(14)) {
+			m_state = 1;
+
+			if (m_worldSpeed < g_eight) {
+				m_worldSpeed = g_eight;
+			}
+
+			m_destEdge = LegoPathController::GetControlEdgeA(15);
+			m_boundary = LegoPathController::GetControlBoundaryA(15);
+		}
+
+		if (m_state == 1) {
+			if (m_userNavFlag) {
+				m_unk0xe4 = 0.5f;
+			}
+
+			v1 = m_destEdge->CCWVertex(*m_boundary);
+			v2 = m_destEdge->CWVertex(*m_boundary);
+			assert(v1 && v2);
+
+			a[0] = (*v1)[0] + ((*v2)[0] - (*v1)[0]) * m_unk0xe4;
+			a[1] = (*v1)[1] + ((*v2)[1] - (*v1)[1]) * m_unk0xe4;
+			a[2] = (*v1)[2] + ((*v2)[2] - (*v1)[2]) * m_unk0xe4;
+
+			m_destEdge->FUN_1002ddc0(*m_boundary, bbb);
+			c.EqualsCross(&bbb, m_boundary->GetUnknown0x14());
+			c.Unitize();
+
+			Mx3DPointFloat worldDirection(m_roi->GetWorldDirection());
+
+			if (!m_userNavFlag) {
+				((Vector2*) &worldDirection)->Mul(-1.0f);
+			}
+
+			if (VTable0x80(m_roi->GetWorldPosition(), worldDirection, a, c)) {
+#ifdef NDEBUG
+				m_unk0x7c = 0;
+				return 0;
+#else
+				assert(0);
+				return -1;
+#endif
+			}
+
+			m_unk0x7c = 0;
+			return 0;
+		}
+		else {
+			return 1;
+		}
+	}
 }
 
-// STUB: LEGO1 0x10081550
+// FUNCTION: LEGO1 0x10081550
 void LegoJetskiRaceActor::VTable0x70(float p_float)
 {
-	// TODO
+	if (m_unk0x0c == 0) {
+		const LegoChar* raceState = VariableTable()->GetVariable(g_raceState);
+		if (stricmp(raceState, g_racing) == 0x0) {
```

Comment:
I think it's clearer to just write it as `!stricmp(raceState, g_racing)` (the way it's done in other places)

Reply:
Funny, I personally find the comparison with zero clearer, but that's a matter of opinion (to me, `!stricmp` feels like it says "does not match" while meaning the opposite). I noticed that the rest of the repo isn't entirely consistent there either, but the exclamation mark pattern is much more prevalent. I'll change it.